### PR TITLE
okteto destroy --all option with log stream

### PR DESCRIPTION
**Pull request by teresaromero**
Resolves [#5113](https://github.com/okteto/app/issues/5113)

Depends on #3249 

Proposed changes:
- New `--all` flag for `okteto destroy` cmd
- When this option is enabled, new flow for `okteto destroy`
- New function `DestroyAll` under `Namespaces` okteto client calling `destroyAllInSpace` mutation
- New function `DestroyAllLogs` under `streamClient` and print function to show the logs of this process

### CODE REVIEW

File: pkg/okteto/stream.go
Comment (ifbyol) [inline pkg/okteto/stream.go:98]:
```suggestion
		// skip when the message is `Done`
```

File: cmd/destroy/destroy.go
Comment (ifbyol) [inline cmd/destroy/destroy.go:198]: I don't think this logic has to be here. As I see, the structure of this (and other commands) are:
* Options/flags validation
* Running some pre configuration (mainly load the context)
* And then, command logic (usually through an struct like `destroyCommand`

In this case, we are including part of the command logic in the place where we are validating flags.

I would keep the logic of validating if the context is an Okteto one, but the rest, I would move it to:
* Within the function `runDestroy`
* Another function in the `destroyCommand` struct and in the line `217` include a conditional like

```
if options.DestroyAll {
  err = c.runDestroyAll(ctx, options)
} else { 
  err = c.runDestroy(ctx, options)
}
```

At the end, this logic belongs to `destroyCommand`.

Also, we wouldn't be tracking the destroy as we are doing in the line 218. Should we track it in this case too in an special way?

File: cmd/destroy/destroy.go
Comment (pchico83) [inline cmd/destroy/destroy.go:169]: @rberrelleza could you review this message?

File: pkg/okteto/stream.go
Comment (pchico83) [inline pkg/okteto/stream.go:90]: what happens with reconnections? could we retry and keep iterating on the events?

**Review by pchico83 — COMMENTED** [review]
@teresaromero could you attach a screenshot of the final output? or a short video?

**Comment by teresaromero** [comment]
> @teresaromero could you attach a screenshot of the final output? or a short video?

i've uploaded a loom video on how is looking with the latest commit of this PR https://www.loom.com/share/0128877bc0784889afad4206d0f27b59

File: cmd/destroy/destroy.go
Comment (teresaromero) [reply cmd/destroy/destroy.go]: I've done this change, also regarding tracking, i think this should be tracked separatly, or at least detect a different event as is a different use?

An other question that i just came up, i added includeVolumes as default true, wondering if we also want to add the flag -v and based on that?

File: pkg/okteto/stream.go
Comment (pchico83) [reply pkg/okteto/stream.go]: @teresaromero that's great!
will events be shown twice after a reconnection? or do we manage to only show events once?

File: cmd/destroy/destroy.go
Comment (ifbyol) [inline cmd/destroy/destroy.go:195]: We are losing the `TrackDestroy`, @pchico83 is it needed to track this action? Should we include some data in the tracking to indicate that the new flag was being used?

File: pkg/okteto/namespace.go
Comment (ifbyol) [inline pkg/okteto/namespace.go:182]: Why always true? `Destroy` operation already receives a boolean to indicate if the volumes has to be destroyed or not. I think we should use the value of that flag. At the end, the user already has an option to delete volumes or not when deleting all from the UI

cc @pchico83 @rlamana

File: pkg/okteto/stream.go
Comment (ifbyol) [reply pkg/okteto/stream.go:106]: I would apply this to avoid a misleading code comment

**Comment by ifbyol** [comment]
@teresaromero I've been testing it and I think we need a more complex change for 2 reasons:
* We rely on the stream logs endpoint/functionality to consider that the command has ended or not. Conceptually, this is wrong. The command should finish when the destroy all operation is finished, we shouldn't rely on the streamed logs.
* Even if the stream finish correctly, we are not even checking the final state of the operation, so we are reporting back that the command was executed successfully, but it is not the case. These are the screenshot of 2 different scenarios where the destroy all failed but we report back a success.

One of the pipeline which had to be explicitly destroyed failed during the deletion as you can see in the logs but the command says it was successfully

Captura de Pantalla 2022-12-05 a las 13 27 31

In this case, the generic `destroy all` failed and the namespace status is `destroyAllFailed` but the command says it was successful (I force this case in my dev environment) 

Captura de Pantalla 2022-12-05 a las 15 06 50


So, I think we should include the logic to monitor the namespace status and stop the command and return the proper response based on:
* If the namespace status goes from `DestroyingAll` to `DestroyAllFailed`, we should return an error
* If the namespace status goes from `DestroyingAll` to `Active`, we should list all configmaps belonging to a dev environment and check if there is any in a `error` or `destroy-error` status. If there is any, the command should return an error. If there isn't any, the command should return a success

File: pkg/okteto/namespace.go
Comment (teresaromero) [reply pkg/okteto/namespace.go:182]: i thought that "all" meant all... so making this default to true will also remove the volumes 🤷🏻‍♀️  
WDYT? no problem in changing it

File: cmd/destroy/destroy.go
Comment (teresaromero) [reply cmd/destroy/destroy.go:195]: is we keep the same event for both destroys we wont see the difference of using `--all` flag, i think we either add more meta to this event or we create a new event for destroyall. WDYT?

File: pkg/okteto/namespace.go
Comment (pchico83) [reply pkg/okteto/namespace.go:182]: agree with @ifbyol. `all` means all the repos and services. we already have `-v` to destroy volumes

File: cmd/destroy/destroy.go
Comment (pchico83) [reply cmd/destroy/destroy.go:195]: let's add an additional field to the `destroy` event. It's what we do for other commands

File: pkg/okteto/namespace.go
Comment (teresaromero) [reply pkg/okteto/namespace.go]: added this value at the mutation

**Comment by ifbyol** [comment]
> > @teresaromero I've been testing it and I think we need a more complex change for 2 reasons:
> > 
> > * We rely on the stream logs endpoint/functionality to consider that the command has ended or not. Conceptually, this is wrong. The command should finish when the destroy all operation is finished, we shouldn't rely on the streamed logs.
> > * Even if the stream finish correctly, we are not even checking the final state of the operation, so we are reporting back that the command was executed successfully, but it is not the case. These are the screenshot of 2 different scenarios where the destroy all failed but we report back a success.
> > 
> > One of the pipeline which had to be explicitly destroyed failed during the deletion as you can see in the logs but the command says it was successfully
> > Captura de Pantalla 2022-12-05 a las 13 27 31
> > In this case, the generic `destroy all` failed and the namespace status is `destroyAllFailed` but the command says it was successful (I force this case in my dev environment)
> > Captura de Pantalla 2022-12-05 a las 15 06 50
> > So, I think we should include the logic to monitor the namespace status and stop the command and return the proper response based on:
> > 
> > * If the namespace status goes from `DestroyingAll` to `DestroyAllFailed`, we should return an error
> > * If the namespace status goes from `DestroyingAll` to `Active`, we should list all configmaps belonging to a dev environment and check if there is any in a `error` or `destroy-error` status. If there is any, the command should return an error. If there isn't any, the command should return a success
> 
> I was looking into it, and as i see without making any changes at the other components i see a solution on calling namespaces list, and check the status of the deleted namespace.
> 
> Other option could be to modify the mutation so it returns the id of the job, and call to see the progress of this?

I would go with checking the namespace status using k8s API directly or even using a watcher in the namespace. Even when the namespace comes back to `Active`, we should check if something is still deployed there or not as it might have failed in a pipeline destroy, in which case, the status goes back to `Active`

File: cmd/destroy/destroy.go
Comment (ifbyol) [inline cmd/destroy/destroy.go:539]:
```suggestion
					// when status is active again check if all resources have been correctly destroyed
```

File: cmd/destroy/destroy.go
Comment (ifbyol) [inline cmd/destroy/destroy.go:550]: Probably checking that `len(cfgList.Items) > 0` is enough (and better). As the destroy all operation should be intended to delete everything, if a configmap is still there, it means that something failed

File: cmd/destroy/destroy.go
Comment (ifbyol) [inline cmd/destroy/destroy.go:548]:
```suggestion
							return fmt.Errorf("namespace destroy all failed: some resources were not destroyed")
```

File: pkg/okteto/stream.go
Comment (teresaromero) [reply pkg/okteto/stream.go]: @pchico83  regarding reconnection, my mistake, what is implemented at the request for subscription is a retry for the initial connection, as this was a leading point to solve cloud issue...

a reconnection scenario where the connection is lost would require more changes. The logs come though the body stream, so we would need to detect if the scan of the body has stoped because a connection failure. the logs are pure text so there is no id or order we can identify though the stream... i think a reconnection strategy would need more refinement, as while the process continues, the reconnection time could lead us to miss some of the logs.

cc: @ifbyol

**Review by ifbyol — APPROVED** [review]
It looks good minor comments to take a look but approved.

I was testing the branch and I think the streaming of logs could be a bit misleading. Check how it look like

Captura de Pantalla 2022-12-15 a las 19 33 36

I think we might print the streamed log in a different color or with some indentation. WDYT @pchico83 @rlamana ? If we want to improve something, I wouldn't block the PR for this, this is already an improvement but maybe we can create a ticket or a follow up PR

File: pkg/okteto/stream.go
Comment (ifbyol) [reply pkg/okteto/stream.go]: Yeah, I wouldn't block this PR for that reconnection. As tere says, we don't have any kind of id, so it would require some extra work or refinement

**Comment by teresaromero** [comment]
> I think we might print the streamed log in a different color or with some indentation. WDYT @pchico83 @rlamana ? If we want to improve something, I wouldn't block the PR for this, this is already an improvement but maybe we can create a ticket or a follow up PR

Formatting this text is done showing the log line with oktetoLog.Println. this was already used when streaming pipeline logs. changing this format would imply more changes, i think we should not block the PR because of this and  if format change is wanted open an issue to follow that enhacement.

File: cmd/destroy/destroy.go
Comment (teresaromero) [reply cmd/destroy/destroy.go]: > we should list all configmaps belonging to a dev environment and check if there is any in a error or destroy-error status. If there is any, the command should return an error. If there isn't any, the command should return a success

i understood that `if there is any` was regarding the configmaps errors. you meant any configmap then?

Comment (ifbyol) [reply cmd/destroy/destroy.go]: No, at that time I was thinking just on "any configmap on that error state". But I mean, If there is any config map (regardless the state), we didn't delete all, right? 

Anyways, we can leave it as it is and change it in the future if we detect any corner case

**Comment by rlamana** [comment]
Agree with @ifbyol, actually in the output he shows I'm unable to distinguish anything, can you show rhe parts that would use indentation?

File: pkg/okteto/stream.go
Comment (pchico83) [reply pkg/okteto/stream.go]: @teresaromero could you create an issue in the product backlog with these details? that would help us to prioritize it as needed

File: cmd/destroy/destroy.go
Comment (jLopezbarb) [inline cmd/destroy/destroy.go:197]: Should we track the destroy here?

Comment (ifbyol) [reply cmd/destroy/destroy.go:197]: It is being tracked in line 207, right? We are adding a new option/flag `DestroyAll`

Comment (jLopezbarb) [reply cmd/destroy/destroy.go:197]: I though it was in the else branch 😅

**Comment by jLopezbarb** [comment]
Tested and code is working!! Please check deepsource before merging it but looks very cool

**Review by jLopezbarb — APPROVED** [review]

File: cmd/destroy/destroy.go
Comment (teresaromero) [reply cmd/destroy/destroy.go]: i think is more accurate that if the cfmap are still there, the job is not success, i will change it

File: cmd/destroy/destroy.go
Comment (teresaromero) [inline cmd/destroy/destroy.go:195]: could we check here the copy of this message if is the correct one? 

cc: @ifbyol @pchico83 @rberrelleza @rlamana

Comment (pchico83) [reply cmd/destroy/destroy.go:195]: @teresaromero instead of failing, could we ignore the logic to destroy the pipelines in this case?

Comment (teresaromero) [reply cmd/destroy/destroy.go:195]: @pchico83 you mean at the destroy-all job that is triggered?

**Comment by teresaromero** [comment]
> 



> Agree with @ifbyol, actually in the output he shows I'm unable to distinguish anything, can you show rhe parts that would use indentation?

@rlamana https://github.com/okteto/app/issues/5113 at the last comment of the issue you can find a demo video where the output is shown. 

i would need to know if there is something needed with format right now that can be done or we block this until there is a format solution.

The video shows the 2 affected "streams" which are right now pipeline destroy, pipeline deploy and destroy --all in this PR

**Comment by rlamana** [comment]
@teresaromero I think this is fine. We might need to rethink how we are showing this in general, but following the same mechanism we already have in other parts, lgtm 👌

**Review by rlamana — APPROVED** [review]

**Pull request merged**
